hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunction.java

```diff
@@ -0,0 +1,209 @@
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.VariExpression;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import org.apache.calcite.sql.SqlJsonQueryEmptyOrErrorBehavior;
+import org.apache.calcite.sql.SqlJsonQueryWrapperBehavior;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+import static com.hazelcast.internal.util.StringUtil.isNullOrEmpty;
+import static com.hazelcast.jet.sql.impl.expression.json.JsonPathUtil.isArrayOrObject;
+
+@SuppressWarnings("checkstyle:MagicNumber")
+public class JsonQueryFunction extends VariExpression<HazelcastJsonValue> implements IdentifiedDataSerializable {
+    private static final Gson SERIALIZER = new Gson();
+    private SqlJsonQueryWrapperBehavior wrapperBehavior;
+    private SqlJsonQueryEmptyOrErrorBehavior onEmpty;
+    private SqlJsonQueryEmptyOrErrorBehavior onError;
+
+    public JsonQueryFunction() { }
+
+    private JsonQueryFunction(Expression<?>[] operands,
+                              SqlJsonQueryWrapperBehavior wrapperBehavior,
+                              SqlJsonQueryEmptyOrErrorBehavior onEmpty,
+                              SqlJsonQueryEmptyOrErrorBehavior onError) {
+        super(operands);
+        this.wrapperBehavior = wrapperBehavior;
+        this.onEmpty = onEmpty;
+        this.onError = onError;
+    }
+
+    public static JsonQueryFunction create(Expression<?> json,
+                                           Expression<?> path,
+                                           SqlJsonQueryWrapperBehavior wrapperBehavior,
+                                           SqlJsonQueryEmptyOrErrorBehavior onEmpty,
+                                           SqlJsonQueryEmptyOrErrorBehavior onError) {
+        final Expression<?>[] operands = new Expression<?>[] {
+                json,
+                path
+        };
+        return new JsonQueryFunction(operands, wrapperBehavior, onEmpty, onError);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return JetSqlSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return JetSqlSerializerHook.JSON_QUERY;
+    }
+
+    @Override
+    public HazelcastJsonValue eval(final Row row, final ExpressionEvalContext context) {
+        final Object operand0 = operands[0].eval(row, context);
+        final String json = operand0 instanceof HazelcastJsonValue
+                ? operand0.toString()
+                : (String) operand0;
+        final String path = (String) operands[1].eval(row, context);
+
+        if (isNullOrEmpty(path)) {
+            return onErrorResponse(onError, QueryException.error("JSON_QUERY path expression is empty"));
+        }
+
+        if (isNullOrEmpty(json)) {
```

Comment:
This check should be right after we evaluate `operand0`. Also in `JsonValueFunction`.

Reply:
L86-88 has to extract the value of the string first though to check if its empty actually. I checked and there shouldn't be any NPEs here, `operand0 instanceof HazelcastJsonValue` will always eval to `false` on `null` value and cast itself doesn't do anything unless its a primitive type.

Reply:
The point is that we normally don't evaluate operand1, if we can return the result based on just operand0. If operand0 is null or empty, we'll return empty response regardless of operand1.